### ISSUE TITLE
Startup error (ZMQError)

### ISSUE THREAD

**Issue by moldcraft** [OpenBazaar/OpenBazaar-Server#305]
Operating System (OS and version): Ubuntu
OpenBazaar version: 
Hardware: Inside docker container

Reproducible: Always 

Steps to reproduce:
1. Install dependecies and git clone
2. Start

    ```
    $ python openbazaard.py start -p 12060 -r 12061 -w 12062 -b 12063
    ________                      __________
    \_____  \ ______   ____   ____\______   \_____  _____________  _____ _______
     /   |   \\____ \_/ __ \ /    \|    |  _/\__  \ \___   /\__  \ \__  \\_  __ \ 
    /    |    \  |_> >  ___/|   |  \    |   \ / __ \_/    /  / __ \_/ __ \|  | \/
    \_______  /   __/ \___  >___|  /______  /(____  /_____ \(____  (____  /__|
            \/|__|        \/     \/       \/      \/      \/     \/     \/
    
    OpenBazaar Server v0.1 starting...
    Generating GUID, this may take a few minutes...
    Exception : No UPnP device discovered
    2016-03-18 10:04:27+0000 [OpenBazaard] [INFO] Finding NAT Type...
    2016-03-18 10:04:35+0000 [OpenBazaard] [INFO] Symmetric NAT on ...:1024
    2016-03-18 10:04:35+0000 [Server] [INFO] querying seed.openbazaar.org:8080 for peers
    2016-03-18 10:04:36+0000 [Server] [INFO] seed.openbazaar.org:8080 returned 50 addresses
    Exception in thread Thread-1:
    Traceback (most recent call last):
      File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
        self.run()
      File "/usr/lib/python2.7/threading.py", line 763, in run
        self.__target(*self.__args, **self.__kwargs)
      File "/observer/keys/keychain.py", line 51, in create_keychain
        callback(self, True)
      File "openbazaard.py", line 140, in start_server
        libbitcoin_client = LibbitcoinClient(LIBBITCOIN_SERVER, log=Logger(service="LibbitcoinClient"))
      File "/usr/local/lib/python2.7/dist-packages/obelisk/client.py", line 68, in __init__
        ClientBase.__init__(self, address)
      File "/usr/local/lib/python2.7/dist-packages/obelisk/zmqbase.py", line 35, in __init__
        self._socket = self.setup(address, public_key)
      File "/usr/local/lib/python2.7/dist-packages/obelisk/zmqbase.py", line 158, in setup
        s = ZmqSocket(self.frame_received, self.zmq_version)
      File "/usr/local/lib/python2.7/dist-packages/obelisk/zmq_fallback.py", line 20, in __init__
        self.public_key, self.secret_key = zmq.curve_keypair()
      File "zmq/backend/cython/utils.pyx", line 60, in zmq.backend.cython.utils.curve_keypair (zmq/backend/cython/utils.c:923)
      File "/usr/local/lib/python2.7/dist-packages/zmq/error.py", line 135, in _check_rc
        raise ZMQError(errno)
    ZMQError: Operation not supported
    
    2016-03-18 10:04:39+0000 [Server] [INFO] bootstrapping with 50 addresses, finding neighbors...
    ```

Observed Behavior: Error, but the server was started successfully

Expected Behavior: Start without any errors

**Comment by moldcraft** [comment]
Also, how to stop it correctly? I hit `Ctrl + C` and I got errors

```
2016-03-18 10:04:44+0000 [OpenBazaard] [INFO] bootstrap complete
^C2016-03-18 10:11:11+0000 [-] Unhandled error in Deferred:
Unhandled error in Deferred:
2016-03-18 10:11:11+0000 [-] Unhandled Error
	Traceback (most recent call last):
	  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/defer.py", line 577, in _runCallbacks
	    current.result = callback(current.result, *args, **kw)
	  File "/observer/dht/network.py", line 195, in initTable
	    d.callback(True)
	  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/defer.py", line 382, in callback
	    self._startRunCallbacks(result)
	  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/defer.py", line 490, in _startRunCallbacks
	    self._runCallbacks()
	--- <exception caught here> ---
	  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/defer.py", line 577, in _runCallbacks
	    current.result = callback(current.result, *args, **kw)
	  File "openbazaard.py", line 75, in on_bootstrap_complete
	    task.LoopingCall(mserver.get_messages, mlistener).start(3600)
	exceptions.NameError: free variable 'mlistener' referenced before assignment in enclosing scope
	
Unhandled Error
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/defer.py", line 577, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/observer/dht/network.py", line 195, in initTable
    d.callback(True)
  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/defer.py", line 382, in callback
    self._startRunCallbacks(result)
  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/defer.py", line 490, in _startRunCallbacks
    self._runCallbacks()
--- <exception caught here> ---
  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/defer.py", line 577, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "openbazaard.py", line 75, in on_bootstrap_complete
    task.LoopingCall(mserver.get_messages, mlistener).start(3600)
exceptions.NameError: free variable 'mlistener' referenced before assignment in enclosing scope
```

**Comment by moldcraft** [comment]
Second (and next times) I try to start the server, it stays blocked on error (Ctrl+C doesn't work, I kill the `python` process with sudo)

```
...
2016-03-18 10:19:11+0000 [Server] [INFO] seed.openbazaar.org:8080 returned 50 addresses
Traceback (most recent call last):
  File "openbazaard.py", line 301, in <module>
    Parser(OpenBazaard('/tmp/openbazaard.pid'))
  File "openbazaard.py", line 225, in __init__
    getattr(self, args.command)()
  File "openbazaard.py", line 263, in start
    int(args.heartbeatport), time.time())
  File "openbazaard.py", line 193, in run
    KeyChain(db, start_server, heartbeat_server)
  File "/observer/keys/keychain.py", line 28, in __init__
    callback(self)
  File "openbazaard.py", line 140, in start_server
    libbitcoin_client = LibbitcoinClient(LIBBITCOIN_SERVER, log=Logger(service="LibbitcoinClient"))
  File "/usr/local/lib/python2.7/dist-packages/obelisk/client.py", line 68, in __init__
    ClientBase.__init__(self, address)
  File "/usr/local/lib/python2.7/dist-packages/obelisk/zmqbase.py", line 35, in __init__
    self._socket = self.setup(address, public_key)
  File "/usr/local/lib/python2.7/dist-packages/obelisk/zmqbase.py", line 158, in setup
    s = ZmqSocket(self.frame_received, self.zmq_version)
  File "/usr/local/lib/python2.7/dist-packages/obelisk/zmq_fallback.py", line 20, in __init__
    self.public_key, self.secret_key = zmq.curve_keypair()
  File "zmq/backend/cython/utils.pyx", line 60, in zmq.backend.cython.utils.curve_keypair (zmq/backend/cython/utils.c:923)
  File "/usr/local/lib/python2.7/dist-packages/zmq/error.py", line 135, in _check_rc
    raise ZMQError(errno)
zmq.error.ZMQError: Operation not supported
^C
```

**Comment by insanity54** [comment]
What version of ubuntu are you using? 14.04?

I think you might need libsodium.

https://download.libsodium.org/doc/installation/index.html

**Comment by moldcraft** [comment]
@insanity54 14.04

I installed successfully libsodium from `.tar.gz` but the error is still there

```
2016-03-24 12:13:09+0000 [Server] [INFO] seed.openbazaar.org:8080 returned 43 addresses
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/observer/keys/keychain.py", line 51, in create_keychain
    callback(self, True)
  File "openbazaard.py", line 138, in start_server
    libbitcoin_client = LibbitcoinClient(LIBBITCOIN_SERVERS_TESTNET, log=Logger(service="LibbitcoinClient"))
  File "/usr/local/lib/python2.7/dist-packages/obelisk/client.py", line 68, in __init__
    ClientBase.__init__(self, address)
  File "/usr/local/lib/python2.7/dist-packages/obelisk/zmqbase.py", line 35, in __init__
    self._socket = self.setup(address, public_key)
  File "/usr/local/lib/python2.7/dist-packages/obelisk/zmqbase.py", line 158, in setup
    s = ZmqSocket(self.frame_received, self.zmq_version)
  File "/usr/local/lib/python2.7/dist-packages/obelisk/zmq_fallback.py", line 20, in __init__
    self.public_key, self.secret_key = zmq.curve_keypair()
  File "zmq/backend/cython/utils.pyx", line 60, in zmq.backend.cython.utils.curve_keypair (zmq/backend/cython/utils.c:923)
  File "/usr/local/lib/python2.7/dist-packages/zmq/error.py", line 135, in _check_rc
    raise ZMQError(errno)
ZMQError: Operation not supported

2016-03-24 12:13:18+0000 [Server] [INFO] bootstrapping with 43 addresses, finding neighbors...
```

Also wen I start the client on my PC (the server is running on a remote machine), I get these errors

```
2016-03-24 12:15:02+0000 [txws.WebSocketFactory] Unhandled Error
	Traceback (most recent call last):
	  File "/usr/local/lib/python2.7/dist-packages/twisted/python/log.py", line 73, in callWithContext
	    return context.call({ILogContext: newCtx}, func, *args, **kw)
	  File "/usr/local/lib/python2.7/dist-packages/twisted/python/context.py", line 118, in callWithContext
	    return self.currentContext().callWithContext(ctx, func, *args, **kw)
	  File "/usr/local/lib/python2.7/dist-packages/twisted/python/context.py", line 81, in callWithContext
	    return func(*args,**kw)
	  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/posixbase.py", line 614, in _doReadOrWrite
	    why = selectable.doRead()
	--- <exception caught here> ---
	  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/tcp.py", line 1069, in doRead
	    protocol.makeConnection(transport)
	  File "/usr/local/lib/python2.7/dist-packages/twisted/protocols/policies.py", line 75, in makeConnection
	    self.wrappedProtocol.makeConnection(self)
	exceptions.AttributeError: 'NoneType' object has no attribute 'makeConnection'
	
Unhandled Error
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/twisted/python/log.py", line 73, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "/usr/local/lib/python2.7/dist-packages/twisted/python/context.py", line 118, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/usr/local/lib/python2.7/dist-packages/twisted/python/context.py", line 81, in callWithContext
    return func(*args,**kw)
  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/posixbase.py", line 614, in _doReadOrWrite
    why = selectable.doRead()
--- <exception caught here> ---
  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/tcp.py", line 1069, in doRead
    protocol.makeConnection(transport)
  File "/usr/local/lib/python2.7/dist-packages/twisted/protocols/policies.py", line 75, in makeConnection
    self.wrappedProtocol.makeConnection(self)
exceptions.AttributeError: 'NoneType' object has no attribute 'makeConnection'
```

The client is loading without showing any errors

![](http://static.md/efb65e476d06e75e6d3f0c3873cb6451.png)

![](https://static.md/8dcd1f6875fb92ffd822c803e47b7bef.png)

**Comment by insanity54** [comment]
I'm not sure if libsodium installed correctly. On your Ubuntu system, do you see a folder `/usr/local/include/sodium`?

Did you run `pip install --upgrade -r requirements.txt` after installing sodium from tarball? I'm not sure, but I was thinking ZMQ might reconfigure during this step to use whatever libsodium it can find.

I've seen WebSocket errors too. One potential issue there is that the OpenBazaar team pushes their commits directly to the `master` banch. In other words, you're getting a bleeding edge development version when you download from github, which may or may not be ready to use. The OpenBazaar version downloadable from openbazaar.org are releases considered ready to use.

Either way, you can play around with checking out different commits. The latest commit may have some work-in-progress changes which have broken some things, so you'd want to use an earlier commit without the breaking changes. I'm using OpenBazaar-Server commit `c5d47a4021f7066960656400faa5e758e79c3885` at the moment. It's from March 17 and with it, I'm not getting the websocket error I saw with the latest commit.

(in the openbazaar server directory)

```
$ git checkout c5d47a4021f7066960656400faa5e758e79c3885
$ python openbazaard.py start -t
```

**Comment by moldcraft** [comment]
> On your Ubuntu system, do you see a folder `/usr/local/include/sodium`?

Yes

> Did you run `pip install --upgrade -r requirements.txt` after installing sodium from tarball?

Yes

----

```
observer@bf7f651cd1ef:/observer$ git checkout c5d47a4021f7066960656400faa5e758e79c3885
Note: checking out 'c5d47a4021f7066960656400faa5e758e79c3885'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b new_branch_name

HEAD is now at c5d47a4... Fix but in btc_price
observer@bf7f651cd1ef:/observer$ exit
exit
root@bf7f651cd1ef:/observer# pip install --upgrade -r requirements.txt
Requirement already up-to-date: protobuf==3.0.0b2 in /usr/local/lib/python2.7/dist-packages (from -r requirements.txt (line 1))
Requirement already up-to-date: Twisted==14.0.2 in /usr/local/lib/python2.7/dist-packages (from -r requirements.txt (line 2))
Requirement already up-to-date: txrudp==0.5.1 in /usr/local/lib/python2.7/dist-packages (from -r requirements.txt (line 3))
Requirement already up-to-date: pystun==0.1.0 in /usr/local/lib/python2.7/dist-packages (from -r requirements.txt (line 4))
Requirement already up-to-date: bitcointools==1.1.44 in /usr/local/lib/python2.7/dist-packages (from -r requirements.txt (line 5))
Requirement already up-to-date: gnupg==2.0.2 in /usr/local/lib/python2.7/dist-packages (from -r requirements.txt (line 6))
Requirement already up-to-date: pynacl==1.0.1 in /usr/local/lib/python2.7/dist-packages (from -r requirements.txt (line 7))
Requirement already up-to-date: txrestapi==0.2 in /usr/local/lib/python2.7/dist-packages (from -r requirements.txt (line 8))
Requirement already up-to-date: txws==0.9.1 in /usr/local/lib/python2.7/dist-packages (from -r requirements.txt (line 9))
Downloading/unpacking python-libbitcoinclient==0.4.0 (from -r requirements.txt (line 10))
  Downloading python-libbitcoinclient-0.4.0.tar.gz (41kB): 41kB downloaded
  Running setup.py (path:/tmp/pip_build_root/python-libbitcoinclient/setup.py) egg_info for package python-libbitcoinclient
    
Requirement already up-to-date: python-bitcoinlib==0.5.0 in /usr/local/lib/python2.7/dist-packages (from -r requirements.txt (line 11))
Requirement already up-to-date: pyopenssl==0.15.1 in /usr/local/lib/python2.7/dist-packages (from -r requirements.txt (line 12))
Requirement already up-to-date: miniupnpc==1.9 in /usr/local/lib/python2.7/dist-packages (from -r requirements.txt (line 13))
Requirement already up-to-date: bleach==1.4.2 in /usr/local/lib/python2.7/dist-packages (from -r requirements.txt (line 14))
Requirement already up-to-date: service_identity in /usr/local/lib/python2.7/dist-packages (from -r requirements.txt (line 15))
Requirement already up-to-date: zope.interface>=3.6.0 in /usr/local/lib/python2.7/dist-packages (from Twisted==14.0.2->-r requirements.txt (line 2))
Requirement already up-to-date: enum34 in /usr/local/lib/python2.7/dist-packages (from txrudp==0.5.1->-r requirements.txt (line 3))
Requirement already up-to-date: psutil>=1.2.1 in /usr/local/lib/python2.7/dist-packages (from gnupg==2.0.2->-r requirements.txt (line 6))
Requirement already up-to-date: six in /usr/local/lib/python2.7/dist-packages (from pynacl==1.0.1->-r requirements.txt (line 7))
Requirement already up-to-date: cffi>=1.1.0 in /usr/local/lib/python2.7/dist-packages (from pynacl==1.0.1->-r requirements.txt (line 7))
Requirement already up-to-date: ecdsa in /usr/local/lib/python2.7/dist-packages (from python-libbitcoinclient==0.4.0->-r requirements.txt (line 10))
Requirement already up-to-date: pyzmq in /usr/local/lib/python2.7/dist-packages (from python-libbitcoinclient==0.4.0->-r requirements.txt (line 10))
Requirement already up-to-date: setuptools in /usr/local/lib/python2.7/dist-packages (from zope.interface>=3.6.0->Twisted==14.0.2->-r requirements.txt (line 2))
Requirement already up-to-date: pycparser in /usr/local/lib/python2.7/dist-packages (from cffi>=1.1.0->pynacl==1.0.1->-r requirements.txt (line 7))
Installing collected packages: python-libbitcoinclient
  Found existing installation: python-libbitcoinclient 0.4.2
    Can't uninstall 'python-libbitcoinclient'. No files were found to uninstall.
  Running setup.py install for python-libbitcoinclient
    
  Could not find .egg-info directory in install record for python-libbitcoinclient==0.4.0 (from -r requirements.txt (line 10))
Successfully installed python-libbitcoinclient
Cleaning up...
root@bf7f651cd1ef:/observer# su observer
observer@bf7f651cd1ef:/observer$ cat start.bash 
python openbazaard.py start -t -a 0.0.0.0 -p 12060 -r 12061 -w 12062 -b 12063
observer@bf7f651cd1ef:/observer$ ./start.bash 
________                      __________
\_____  \ ______   ____   ____\______   \_____  _____________  _____ _______
 /   |   \\____ \_/ __ \ /    \|    |  _/\__  \ \___   /\__  \ \__  \\_  __ \ 
/    |    \  |_> >  ___/|   |  \    |   \ / __ \_/    /  / __ \_/ __ \|  | \/
\_______  /   __/ \___  >___|  /______  /(____  /_____ \(____  (____  /__|
        \/|__|        \/     \/       \/      \/      \/     \/     \/

OpenBazaar Server v0.1 starting...
Exception : No UPnP device discovered
2016-03-24 14:06:33+0000 [OpenBazaard] [INFO] Finding NAT Type...
2016-03-24 14:06:42+0000 [OpenBazaard] [INFO] Symmetric NAT on 93.113.113.50:1024
Traceback (most recent call last):
  File "openbazaard.py", line 301, in <module>
    Parser(OpenBazaard('/tmp/openbazaard.pid'))
  File "openbazaard.py", line 225, in __init__
    getattr(self, args.command)()
  File "openbazaard.py", line 263, in start
    int(args.heartbeatport), time.time())
  File "openbazaard.py", line 193, in run
    KeyChain(db, start_server, heartbeat_server)
  File "/observer/keys/keychain.py", line 28, in __init__
    callback(self)
  File "openbazaard.py", line 138, in start_server
    libbitcoin_client = LibbitcoinClient(LIBBITCOIN_SERVER_TESTNET, log=Logger(service="LibbitcoinClient"))
  File "/usr/local/lib/python2.7/dist-packages/obelisk/client.py", line 68, in __init__
    ClientBase.__init__(self, address)
  File "/usr/local/lib/python2.7/dist-packages/obelisk/zmqbase.py", line 35, in __init__
    self._socket = self.setup(address, public_key)
  File "/usr/local/lib/python2.7/dist-packages/obelisk/zmqbase.py", line 158, in setup
    s = ZmqSocket(self.frame_received, self.zmq_version)
  File "/usr/local/lib/python2.7/dist-packages/obelisk/zmq_fallback.py", line 20, in __init__
    self.public_key, self.secret_key = zmq.curve_keypair()
  File "zmq/backend/cython/utils.pyx", line 60, in zmq.backend.cython.utils.curve_keypair (zmq/backend/cython/utils.c:923)
  File "/usr/local/lib/python2.7/dist-packages/zmq/error.py", line 135, in _check_rc
    raise ZMQError(errno)
zmq.error.ZMQError: Operation not supported
```

**Comment by drwasho** [comment]
Is this still a problem @moldcraft ?

**Comment by moldcraft** [comment]
Yes

```
observer@bf7f651cd1ef:/observer$ git pull origin
Updating d690c66..617fa24
Fast-forward
 api/restapi.py         |  42 +++++++++++++++++++++++++----------------
 api/ws.py              |  33 ++++++++++++++++++++++++++------
 db/datastore.py        |  37 +++++++++++++++++++++++++++++++++++-
 market/contracts.py    |  59 ++++++++++++++++++++++++++++++++++++++++-----------------
 market/network.py      | 135 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++----------------------------------------------------
 market/transactions.py |  12 ++++++++++++
 net/wireprotocol.py    |   4 ----
 openbazaard.py         |   4 +++-
 8 files changed, 227 insertions(+), 99 deletions(-)
observer@bf7f651cd1ef:/observer$ ./start.bash       
________                      __________
\_____  \ ______   ____   ____\______   \_____  _____________  _____ _______
 /   |   \\____ \_/ __ \ /    \|    |  _/\__  \ \___   /\__  \ \__  \\_  __ \ 
/    |    \  |_> >  ___/|   |  \    |   \ / __ \_/    /  / __ \_/ __ \|  | \/
\_______  /   __/ \___  >___|  /______  /(____  /_____ \(____  (____  /__|
        \/|__|        \/     \/       \/      \/      \/     \/     \/

OpenBazaar Server v0.1 starting...
Exception : No UPnP device discovered
2016-03-29 15:51:50+0000 [OpenBazaard] [INFO] Finding NAT Type...
2016-03-29 15:51:59+0000 [OpenBazaard] [INFO] Symmetric NAT on 93.113.113.50:1024
Traceback (most recent call last):
  File "openbazaard.py", line 303, in <module>
    Parser(OpenBazaard('/tmp/openbazaard.pid'))
  File "openbazaard.py", line 227, in __init__
    getattr(self, args.command)()
  File "openbazaard.py", line 265, in start
    int(args.heartbeatport), time.time())
  File "openbazaard.py", line 195, in run
    KeyChain(db, start_server, heartbeat_server)
  File "/observer/keys/keychain.py", line 28, in __init__
    callback(self)
  File "openbazaard.py", line 140, in start_server
    libbitcoin_client = LibbitcoinClient(LIBBITCOIN_SERVERS_TESTNET, log=Logger(service="LibbitcoinClient"))
  File "/usr/local/lib/python2.7/dist-packages/obelisk/client.py", line 68, in __init__
    ClientBase.__init__(self, address)
  File "/usr/local/lib/python2.7/dist-packages/obelisk/zmqbase.py", line 35, in __init__
    self._socket = self.setup(address, public_key)
  File "/usr/local/lib/python2.7/dist-packages/obelisk/zmqbase.py", line 158, in setup
    s = ZmqSocket(self.frame_received, self.zmq_version)
  File "/usr/local/lib/python2.7/dist-packages/obelisk/zmq_fallback.py", line 20, in __init__
    self.public_key, self.secret_key = zmq.curve_keypair()
  File "zmq/backend/cython/utils.pyx", line 60, in zmq.backend.cython.utils.curve_keypair (zmq/backend/cython/utils.c:923)
  File "/usr/local/lib/python2.7/dist-packages/zmq/error.py", line 135, in _check_rc
    raise ZMQError(errno)
zmq.error.ZMQError: Operation not supported
```

**Comment by moldcraft** [comment]
I can give you ssh access to my docker container `moldcraft[at]email.com`

**Comment by moldcraft** [comment]
Thank you for https://openbazaar.slack.com/archives/openbazaar-server/p1459546668000827

```
...
PASS: tests/test_spec_dealer
PASS: tests/test_spec_router
PASS: tests/test_spec_pushpull
./config/test-driver: line 107: 24949 Aborted                 (core dumped) "$@" > $log_file 2>&1
XFAIL: tests/test_req_correlate
./config/test-driver: line 107: 24978 Aborted                 (core dumped) "$@" > $log_file 2>&1
XFAIL: tests/test_req_relaxed
PASS: tests/test_conflate
PASS: tests/test_inproc_connect
PASS: tests/test_issue_566
PASS: tests/test_proxy
PASS: tests/test_proxy_single_socket
PASS: tests/test_proxy_terminate
PASS: tests/test_getsockopt_memset
PASS: tests/test_setsockopt
./config/test-driver: line 107: 25404 Aborted                 (core dumped) "$@" > $log_file 2>&1
FAIL: tests/test_many_sockets
PASS: tests/test_ipc_wildcard
PASS: tests/test_diffserv
PASS: tests/test_connect_rid
...
```


```
...
PASS: tests/test_abstract_ipc
make[4]: Entering directory `/observer/libzmq'
Making all in doc
make[5]: Entering directory `/observer/libzmq/doc'
make[5]: Nothing to be done for `all'.
make[5]: Leaving directory `/observer/libzmq/doc'
make[5]: Entering directory `/observer/libzmq'
make[5]: Leaving directory `/observer/libzmq'
make[4]: Leaving directory `/observer/libzmq'
============================================================================
Testsuite summary for zeromq 4.2.0
============================================================================
# TOTAL: 79
# PASS:  76
# SKIP:  0
# XFAIL: 2
# FAIL:  1
# XPASS: 0
# ERROR: 0
============================================================================
See ./test-suite.log
Please report to zeromq-dev@lists.zeromq.org
============================================================================
make[3]: *** [test-suite.log] Error 1
make[3]: Leaving directory `/observer/libzmq'
make[2]: *** [check-TESTS] Error 2
make[2]: Leaving directory `/observer/libzmq'
make[1]: *** [check-am] Error 2
make[1]: Leaving directory `/observer/libzmq'
make: *** [check-recursive] Error 1
```

But still doesn't work

```
observer@bf7f651cd1ef:/observer$ cat start.bash 
python openbazaard.py start -t -a 0.0.0.0 -p 12060 -r 12061 -w 12062 -b 12063
observer@bf7f651cd1ef:/observer$ ./start.bash 
________                      __________
\_____  \ ______   ____   ____\______   \_____  _____________  _____ _______
 /   |   \\____ \_/ __ \ /    \|    |  _/\__  \ \___   /\__  \ \__  \\_  __ \ 
/    |    \  |_> >  ___/|   |  \    |   \ / __ \_/    /  / __ \_/ __ \|  | \/
\_______  /   __/ \___  >___|  /______  /(____  /_____ \(____  (____  /__|
        \/|__|        \/     \/       \/      \/      \/     \/     \/

OpenBazaar Server v0.1 starting...
Exception : No UPnP device discovered
2016-04-02 12:05:43+0000 [OpenBazaard] [INFO] Finding NAT Type...
2016-04-02 12:05:51+0000 [OpenBazaard] [INFO] Symmetric NAT on ############:1024
Traceback (most recent call last):
  File "openbazaard.py", line 304, in <module>
    Parser(OpenBazaard('/tmp/openbazaard.pid'))
  File "openbazaard.py", line 228, in __init__
    getattr(self, args.command)()
  File "openbazaard.py", line 266, in start
    int(args.heartbeatport), time.time())
  File "openbazaard.py", line 196, in run
    KeyChain(db, start_server, heartbeat_server)
  File "/observer/keys/keychain.py", line 28, in __init__
    callback(self)
  File "openbazaard.py", line 141, in start_server
    libbitcoin_client = LibbitcoinClient(LIBBITCOIN_SERVERS_TESTNET, log=Logger(service="LibbitcoinClient"))
  File "/usr/local/lib/python2.7/dist-packages/obelisk/client.py", line 68, in __init__
    ClientBase.__init__(self, address)
  File "/usr/local/lib/python2.7/dist-packages/obelisk/zmqbase.py", line 35, in __init__
    self._socket = self.setup(address, public_key)
  File "/usr/local/lib/python2.7/dist-packages/obelisk/zmqbase.py", line 158, in setup
    s = ZmqSocket(self.frame_received, self.zmq_version)
  File "/usr/local/lib/python2.7/dist-packages/obelisk/zmq_fallback.py", line 20, in __init__
    self.public_key, self.secret_key = zmq.curve_keypair()
  File "zmq/backend/cython/utils.pyx", line 60, in zmq.backend.cython.utils.curve_keypair (zmq/backend/cython/utils.c:923)
  File "/usr/local/lib/python2.7/dist-packages/zmq/error.py", line 135, in _check_rc
    raise ZMQError(errno)
zmq.error.ZMQError: Operation not supported
```

**Comment by hoffmabc** [comment]
Did you pip uninstall pyzmq and reinstall it?

**Comment by moldcraft** [comment]
```
pip uninstall pyzmq
Can't uninstall 'pyzmq'. No files were found to uninstall.
```

**Comment by moldcraft** [comment]
```
# pip install -r requirements.txt
....
Requirement already satisfied (use --upgrade to upgrade): pyzmq in /usr/local/lib/python2.7/dist-packages (from python-libbitcoinclient==0.4.2->-r requirements.txt (line 10))
```

**Comment by hoffmabc** [comment]
Try 
```
pip install --upgrade pip
```
and then remove and reinstall it.

**Comment by moldcraft** [comment]
Same https://github.com/OpenBazaar/OpenBazaar-Server/issues/305#issuecomment-204722746

@drwasho Has the access details to my server, I can also send them to you. I think it will be much faster if you will inspect the problem yourself instead of telling me which command to execute.

I will send you the details on email `brianchoffman[at]gmail.com`.

**Comment by noflag** [comment]
same issue, tried everything in the thread

**Comment by noflag** [comment]
Works for me with the latest pull

**Comment by moldcraft** [comment]
I can't tell the same

```
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/observer/keys/keychain.py", line 51, in create_keychain
    callback(self, True)
  File "openbazaard.py", line 143, in start_server
    libbitcoin_client = LibbitcoinClient(LIBBITCOIN_SERVERS, log=Logger(service="LibbitcoinClient"))
  File "/usr/local/lib/python2.7/dist-packages/obelisk/client.py", line 68, in __init__
    ClientBase.__init__(self, address)
  File "/usr/local/lib/python2.7/dist-packages/obelisk/zmqbase.py", line 35, in __init__
    self._socket = self.setup(address, public_key)
  File "/usr/local/lib/python2.7/dist-packages/obelisk/zmqbase.py", line 159, in setup
    s.connect(address, public_key)
  File "/usr/local/lib/python2.7/dist-packages/obelisk/zmq_fallback.py", line 28, in connect
    self._socket.connect(address)
  File "zmq/backend/cython/socket.pyx", line 509, in zmq.backend.cython.socket.Socket.connect (zmq/backend/cython/socket.c:5307)
TypeError: expected str, got: [('tcp://libbitcoin1.openbazaar.org:9091', 'ptCDh+i&@k3CYEa)(tC=j9$yN/eD&}@LJojDUGo7'), ('tcp://libbitcoin3.openbazaar.org:9091', 'pco)6}^g!]5b7#byrx>[=MCI#W?DCF}K3FoS6nTu')]
```

**Comment by insanity54** [comment]
Following my own instructions, I wasn't able to run on Ubuntu 12.04 due to the error-

    zmq.error.ZMQError: Operation not supported

I'm not a python developer so I don't know too much about the technical aspects of python virtualenv, but I'm having much more success by running openbaazard.py in a virtualenv. I'm assuming this is because the libsodium or ZMQ at the system level has conflicts with OpenBazaar code, but pynacl installed by pip does not? Maybe someone in the know could comment on this.

This results in no server errors for me in 12.04 and 14.04--

make a python virtual environment 

    virtualenv observer

activate the environment

    source observer/bin/activate

install dependencies inside the virtualenv

    pip install -r requirements.txt

start the server inside the virtualenv

    python openbazaard.py start -a [my_client_ip_address]

**Comment by moldcraft** [comment]
@drwasho Finally! :tada:
